Gems/Atom/RHI/Code/Source/RHI/MultiDeviceQueryPool.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RHI/Factory.h>
+#include <Atom/RHI/MultiDeviceQuery.h>
+#include <Atom/RHI/MultiDeviceQueryPool.h>
+#include <Atom/RHI/RHISystemInterface.h>
+
+#include <AzCore/std/parallel/lock.h>
+
+namespace AZ::RHI
+{
+    ResultCode MultiDeviceQueryPool::Init(MultiDevice::DeviceMask deviceMask, const QueryPoolDescriptor& descriptor)
+    {
+        if (Validation::IsEnabled())
+        {
+            if (!descriptor.m_queriesCount)
+            {
+                AZ_Error("RHI", false, "MultiDeviceQueryPool size can't be zero");
+                return ResultCode::InvalidArgument;
+            }
+
+            if (descriptor.m_type == QueryType::PipelineStatistics && descriptor.m_pipelineStatisticsMask == PipelineStatisticsFlags::None)
+            {
+                AZ_Error("RHI", false, "Missing pipeline statistics flags");
+                return ResultCode::InvalidArgument;
+            }
+
+            if (descriptor.m_type != QueryType::PipelineStatistics && descriptor.m_pipelineStatisticsMask != PipelineStatisticsFlags::None)
+            {
+                AZ_Warning(
+                    "RHI",
+                    false,
+                    "Pipeline statistics flags are only valid for PipelineStatistics pools. Ignoring m_pipelineStatisticsMask");
+            }
```

Comment:
Since we want to keep the functionality of single-device objects intact, such that Gems can decide to directly use single-device resources as well, we would like to keep the validation on the single-device level, the question maybe if we would want to remove some of the validation on the multi-device level (there is already less, but somethings are checked multiple times, that is true).

Reply:
Eventually (after renaming) RPI will (should?) not be able to access singe device rhi objects directly in which case we could consolidate code like this. Either it resides at MD level or single device level but maybe not both.

Reply:
What @martinwinter-huawei means is that it should still be possible to use the single device RHI classes directly if so desired by a gem, when doing really low level stuff for example. There it would be nice to have the validation still being done.
The reason why we duplicated it in the multi device class is to have it error early and not by the time the call is forwarded to the first single device object. Of course this duplication means a little more code maintenance. If you prefer we can remove all duplicated validation from the multi device classes and have it only done in the single device ones. I would not do it the other way around though. Even though this means that the validation is less performant, since it is checked multiple times (for every device), it is not used in release mode anyway.

Reply:
I guess the question here is do we want higher level code to access SD objects? My initial response would be **no** as the new Framegraph api should only support adding MD objects. Maybe there maybe a use case out there where we may want to expose SD objects outside of RHI but unless it is needed (without a better way around it) we should try to refrain from doing that. If we stick to this policy then we could remove validation from SD objects and only keep it in MD versions.

Reply:
That's true. If we want to restrict access to SD objects completely, we could do that. In this case I would do these changes later though, when we are at the point, that only MD objects are used by the frame graph, since for now SD objects are used and the validation should still happen. I guess the question for now is, if we should keep the validation in the MD objects or not.

Reply:
yeah. I am okay with these changes atm as removal of validation can come when the final switch takes place.